### PR TITLE
Implement `Crystal::Loader` for MSVC

### PR DESCRIPTION
**Pull request by HertzDevil**
This PR adds an implementation of `Crystal::Loader` that imitates the MSVC linker. The end goal here is to get the interpreter compiling on Windows. The loader can be used stand-alone:

```crystal
require "compiler/crystal/loader"

# these are obtained from running a blank source with `crystal run file.cr --verbose`
# and taking everything after `/link`
args = [
  %q(/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\atlmfc\lib\x64),
  %q(/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\lib\x64),
  %q(/LIBPATH:C:\Program Files (x86)\Windows Kits\10\Lib\10.0.19041.0\ucrt\x64),
  %q(/LIBPATH:C:\Program Files (x86)\Windows Kits\10\Lib\10.0.19041.0\um\x64),
  %q(/DEBUG:FULL),
  %q(/PDBALTPATH:%_PDB%),
  %q(/INCREMENTAL:NO),
  %q(/STACK:0x800000),
  %q(/LIBPATH:C:\crystal\lib),
  %q(pcre.lib),
  %q(gc.lib),
  %q(/ENTRY:wmainCRTStartup),
  %q(iconv.lib),
  %q(advapi32.lib),
  %q(Kernel32.lib),
  %q(shell32.lib),
  %q(ole32.lib),
  %q(WS2_32.lib),
  %q(legacy_stdio_definitions.lib),
  %q(DbgHelp.lib),
  %q(libcmt.lib),
]

loader = Crystal::Loader.parse(args) # => #<Crystal::Loader:0x2e0746c0dc0 ...>
loader.find_symbol("GetProcAddress") # => Pointer(Void)@0x7fff07f3aec0
```

It implements a minimal COFF archive reader to extract the DLLs imported by a `.lib` file. Note that a single library may import zero DLLs, such as static libraries, or multiple, such as `ucrt.lib`.

We have the opportunity to customize the whole search order by only passing absolute paths to `LibC.LoadLibraryEx`. Here the loader prefers DLLs stored alongside the import libraries, then uses the [standard DLL search order](https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order); on my setup this is necessary to select the correct `libiconv.dll`. Whatever we do, this search order should eventually coincide with the one for binaries compiled with load-time dynamic linking.

Related: #11555

### CODE REVIEW

File: src/crystal/system/win32/library_archive.cr [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      reader = COFFReader.new(f)
+      reader.run
+      reader.dlls
+    end
+  end
+
+  # Minimal implementation of a Microsoft COFF archive reader. All unused fields
+  # are ignored.
+  private struct COFFReader
+    getter dlls = Set(String).new
```
Comment (straight-shoota) [inline src/crystal/system/win32/library_archive.cr:18]: Why did you choose `Set` for this? There is an *order* to the filenames so I would expect to preserve that. We can still remove duplicates, of course.

Comment (HertzDevil) [reply src/crystal/system/win32/library_archive.cr:18]: `Set` preserves insertion order, so it would be the same as using an `Array` and calling `#uniq!`, wouldn't it? Or is there some other order that should be observed here?

Comment (asterite) [reply src/crystal/system/win32/library_archive.cr:18]: It's not incidental. The concept of a set guarantees no order. The specific implementation of does. We are not going to change that, ever, as it would be a breaking change. Using a Set is fine here.

Comment (asterite) [reply src/crystal/system/win32/library_archive.cr:18]: Another way to say this: I really like how Set and Hash in Ruby preserve the order. That's why it was done in the same way in Crystal. It's a property of Ruby and Crystal Hash and Set and I'd like to keep it like that. It's efficient, and it also gives a guarantee about the order. There's absolutely no reason to change this, or to hide the fact that this works the way it is.

**Review by oprypin — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/loader.cr [hunk 2]
```diff
@@ -70,35 +78,21 @@ class Crystal::Loader
...
 
-  def load_library?(libname : String) : Handle?
+  def load_library?(libname : String) : Bool
     if ::Path::SEPARATORS.any? { |separator| libname.includes?(separator) }
-      return load_file(::Path[libname].expand)
-    end
-
-    each_library_path(libname) do |library_path|
-      handle = load_file?(library_path)
-      return handle if handle
```
Comment (asterite) [inline src/compiler/crystal/loader.cr:84]: Why was this removed? 🤔
